dagger-core/src/test/java/io/odpf/dagger/core/sink/bigquery/BigquerySinkWriterTest.java

```diff
@@ -0,0 +1,92 @@
+        Mockito.when(sink.pushToSink(Mockito.anyList())).thenReturn(response);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        Mockito.verify(sink, Mockito.times(2)).pushToSink(Mockito.anyList());
+        Mockito.verify(response, Mockito.times(2)).hasErrors();
+    }
+
+    @Test
+    public void shouldNotWriteIfCurrentSizeIsLessThanTheBatchSize() throws IOException {
+        ProtoSerializer protoSerializer = Mockito.mock(ProtoSerializer.class);
+        OdpfSink sink = Mockito.mock(OdpfSink.class);
+        BigquerySinkWriter bigquerySinkWriter = new BigquerySinkWriter(protoSerializer, sink, 10);
+        Row row = new Row(1);
+        row.setField(0, "some field");
+        Mockito.when(protoSerializer.serializeKey(row)).thenReturn("test".getBytes());
+        Mockito.when(protoSerializer.serializeValue(row)).thenReturn("testMessage".getBytes());
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        Mockito.verify(sink, Mockito.times(0)).pushToSink(Mockito.anyList());
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenSinkResponseHasErrors() throws IOException {
+        ProtoSerializer protoSerializer = Mockito.mock(ProtoSerializer.class);
+        OdpfSink sink = Mockito.mock(OdpfSink.class);
+        BigquerySinkWriter bigquerySinkWriter = new BigquerySinkWriter(protoSerializer, sink, 3);
+        Row row = new Row(1);
+        row.setField(0, "some field");
+        Mockito.when(protoSerializer.serializeKey(row)).thenReturn("test".getBytes());
+        Mockito.when(protoSerializer.serializeValue(row)).thenReturn("testMessage".getBytes());
+        OdpfSinkResponse response = Mockito.mock(OdpfSinkResponse.class);
+        Mockito.when(response.hasErrors()).thenReturn(true);
+        Mockito.when(response.getErrors()).thenReturn(new HashMap<Long, ErrorInfo>() {{
+            put(1L, new ErrorInfo(new Exception(), ErrorType.DESERIALIZATION_ERROR));
+            put(2L, new ErrorInfo(new Exception(), ErrorType.SINK_4XX_ERROR));
+        }});
+        Mockito.when(sink.pushToSink(Mockito.anyList())).thenReturn(response);
+        bigquerySinkWriter.write(row, null);
+        bigquerySinkWriter.write(row, null);
+        IOException ioException = Assert.assertThrows(IOException.class, () -> bigquerySinkWriter.write(row, null));
+        Assert.assertEquals("Failed to push 2 records to BigquerySink", ioException.getMessage());
+        Mockito.verify(sink, Mockito.times(1)).pushToSink(Mockito.anyList());
```

Comment:
how did this call happen? our batch size is 3, and we only pushed 2 records

Reply:
The third call is in line 78.